Add Swagger document for the Test Monitor service

BKnight760 opened the pull request:
[x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/systemlink-OpenAPI-documents/blob/master/CONTRIBUTING.md).

### What does this Pull Request accomplish?

This pull request adds a Swagger document for the Test Monitor service. This will be the master copy of the Test Monitor service's Swagger document, for use in SystemLink Server. This document is used as the source for the webservice interface which is implemented by the Test Monitor service.

### Why should this Pull Request be merged?

This pull request gives higher visibility to the Test Monitor HTTP API, and ensures that the implentation matches the fully documented options for the webservice.

### What testing has been done?

Existing automated tests have been run against the swagger interface, and all tests pass.

BKnight760 committed:
Create nitestmonitor.yml

BKnight760 committed:
Add the id as required in the test result response.

pvallone commented on test-monitor/nitestmonitor.yml:
```diff
@@ -0,0 +1,1277 @@
+  - cookieAuth: []
+x-ni-routing-key: Skyline.TestMonitor
+definitions:
+  Error:
```
Jenna suggested some changes to my Error definition, so you should probably try to incorporate those, as these should be pretty similar. At some point in the future we could look into single-sourcing the error type, so we don't all have a copy of it in our docs. I don't think it needs to happen now, given that we're trying to release.

pvallone commented on test-monitor/nitestmonitor.yml:
```diff
@@ -0,0 +1,1277 @@
+      x-ni-operation: queryResults
+      x-ni-privilege: Read
+      summary: Queries test results
+      description: Queries test results
```
If we think it isn't intuitive, we could document that an empty body returns everything.

pvallone commented on test-monitor/nitestmonitor.yml:
```diff
@@ -0,0 +1,1277 @@
+      x-ni-request-all-privileges: true
+      summary: API information
+      description: Returns information about API versions and available operations.
+      tags: [versioning]
```
Do you think we should try to be consistent with respect to how we order these properties across documents? In the message document I define tags, summary, description, operationId, and then responses, in that order. It might not be a big deal if we're at least consistent within each individual document.

pvallone commented on test-monitor/nitestmonitor.yml:
```diff
@@ -0,0 +1,1277 @@
+          $ref: '#/responses/Unauthorized'
+        default:
+          $ref: '#/responses/Error'
+  /v1/ping:
```
Have we shipped documentation for this route to customers? If we haven't, I would think we can just leave this route undocumented, as it's deprecated.

spanglerco replied:
We are using codegen for this service, so it has to be documented unless we remove it from the service (which we probably should anyway).

jjaco2 replied:
Should we note it's deprecated outside of the tags section, like in the description?

BKnight760 replied:
I've marked it deprecated in the description, although I agree we should make any required changes to the UI and drop it from the API entirely.  I'll do that as a separate change since it requires a backend code change.

pvallone reviewed (dismissed):
Light review given Paul and Alex reviewed this already

jjaco2 commented on test-monitor/nitestmonitor.yml:
```diff
@@ -0,0 +1,1277 @@
+    description: Contains error information
+    type: object
+    properties:
+      name:
```
Non-essential but worth noting for consistency in the future: drop leading articles like "The" and "A" throughout all  first sentences of descriptions.

jjaco2 commented on test-monitor/nitestmonitor.yml:
```diff
@@ -0,0 +1,1277 @@
+    properties:
+      operations:
+        description: >-
+          Available operations in the v1 version of the API:
```
Could we remove "The ability to" in all the items below, or am I misunderstanding something?

BKnight760 replied:
This one happens to be odd, because every operation is explicitly identified here.  It may make sense to combine these into more general operations.  Then the explanation of specific requests will make sense.  Since that requires a backend service change, I'll defer to doing that in a separate change.

jjaco2 commented on test-monitor/nitestmonitor.yml:
```diff
@@ -0,0 +1,1277 @@
+      - statusType
+    properties:
+      statusType:
+        description: The status type
```
Do we need to specify this is an enum like we do for ResultField?

jjaco2 commented on test-monitor/nitestmonitor.yml:
```diff
@@ -0,0 +1,1277 @@
+      - DATA_MODEL
+      - DATA
+  ResultSortDefinitionObject:
+    description: A result sort definition object
```
What does this mean? I get that this isn't the most essential part of this doc, but if we're not expecting the user to necessarily have experience with HTTP APIs, and this isn't really an industry term, it could be worth rephrasing in a way that doesn't repeat the object name.

jjaco2 commented on test-monitor/nitestmonitor.yml:
```diff
@@ -0,0 +1,1277 @@
+      status:
+        $ref: '#/definitions/StatusObject'
+      startedAt:
+        description: An ISO-8601 formatted timestamp indicating when the result was started
```
> indicating when the result was started

Maybe "indicating the start of the timeframe for the result" or "indicating when the result began"

jjaco2 commented on test-monitor/nitestmonitor.yml:
```diff
@@ -0,0 +1,1277 @@
+        example:
+          key1: value1
+      fileIds:
+        description: A list of attached file ids
```
A list of file ids attached to the result

jjaco2 commented on test-monitor/nitestmonitor.yml:
```diff
@@ -0,0 +1,1277 @@
+          type: string
+        example: []
+  TestResultUpdateRequestObject:
+    description: A test result update request object
```
Same note as for ResultSortDefinitionObject, and really any object descriptions that match the name verbatim: is there a way to add value to users who might not already be familiar with this domain by rephrasing the name?

Maybe "Object representing test result information for [however this object gets used, I'm not sure]"

Same goes for TestResultQueryObject, too

jjaco2 commented on test-monitor/nitestmonitor.yml:
```diff
@@ -0,0 +1,1277 @@
+        items:
+          $ref: '#/definitions/TimeQueryObject'
+      startedAtQuery:
+        description: A time query object
```
Since this is the same as the previous description, is there a way to differentiate the two?

jjaco2 commented on test-monitor/nitestmonitor.yml:
```diff
@@ -0,0 +1,1277 @@
+        example:
+          - 5982002b5d67371430c80e73
+      statuses:
+        description: An array of resulkt statuses
```
> resulkt

jjaco2 commented on test-monitor/nitestmonitor.yml:
```diff
@@ -0,0 +1,1277 @@
+        items:
+          $ref: '#/definitions/ResultSortDefinitionObject'
+  StepIdResultIdPair:
+    description: A pair of step id and result id
```
Step id and result id

jjaco2 commented on test-monitor/nitestmonitor.yml:
```diff
@@ -0,0 +1,1277 @@
+        items:
+          $ref: '#/definitions/StepIdResultIdPair'
+  TestStepQueryObject:
+    description: A test step query object
```
Maybe "Object for querying a test step"

jjaco2 commented on test-monitor/nitestmonitor.yml:
```diff
@@ -0,0 +1,1277 @@
+      data:
+        $ref: '#/definitions/StepDataObject'
+      dataModel:
+        description: The step data model
```
Data model for the step

jjaco2 commented on test-monitor/nitestmonitor.yml:
```diff
@@ -0,0 +1,1277 @@
+        items:
+          $ref: '#/definitions/TestStepRequestObject'
+  TestStepResponseObject:
+    description: A test step response object
```
Same note about adding value for less HTTP API-savvy users by rephrasing and/or adding (very brief) context

jjaco2 commented on test-monitor/nitestmonitor.yml:
```diff
@@ -0,0 +1,1277 @@
+          items:
+            $ref: '#/definitions/TestResultResponseObject'
+        totalCount:
+          description: The total count of matching test results
```
> count of

number of

jjaco2 commented on test-monitor/nitestmonitor.yml:
```diff
@@ -0,0 +1,1277 @@
+          items:
+            $ref: '#/definitions/TestResultResponseObject'
+        totalCount:
+          description: The total count of matching test results
```
Same note as before (count -> number) Not sure if single-sourced

jjaco2 commented on test-monitor/nitestmonitor.yml:
```diff
@@ -0,0 +1,1277 @@
+      type: object
+      properties:
+        steps:
+          description: An array of test step id result id paris that were deleted
```
> paris

jjaco2 commented on test-monitor/nitestmonitor.yml:
```diff
@@ -0,0 +1,1277 @@
+          items:
+            $ref: '#/definitions/StepIdResultIdPair'
+        failed:
+          description: An array of test step id result id paris that failed to delete
```
> paris

jjaco2 commented on test-monitor/nitestmonitor.yml:
```diff
@@ -0,0 +1,1277 @@
+      x-ni-operation: updateResults
+      x-ni-privilege: Write
+      summary: Update test results
+      description: Update test results
```
Updates existing test results

Alternatively, you could delete the summary from all these routes, since they're usually the same as the description, and IIRC, @pvallone only used descriptions.

jjaco2 commented on test-monitor/nitestmonitor.yml:
```diff
@@ -0,0 +1,1277 @@
+      x-ni-operation: createResults
+      x-ni-privilege: Write
+      summary: Create test results
+      description: Create test results
```
Creates new test results

(or cut the summary)
(this note applies to all the summaries in this section)

jjaco2 commented on test-monitor/nitestmonitor.yml:
```diff
@@ -0,0 +1,1277 @@
+          required: true
+        - in: query
+          name: deleteSteps
+          description: Delete the test steps associated with the test result
```
Indicates whether to delete the test steps associated with the test result

jjaco2 commented on test-monitor/nitestmonitor.yml:
```diff
@@ -0,0 +1,1277 @@
+                  type: string
+                  example: 1f05c3c8-7a8c-4ebb-a87f-7a60d5dd8938
+              deleteSteps:
+                description: Delete the test steps associated with the test result
```
"Indicates whether to...."

jjaco2 reviewed (commented):
Mostly minor things plus some opportunities to briefly clarify some concepts/terminology for more novice users.

BKnight760 committed:
Address review comments to descriptions

This handles a large portion of the review comments.  I may still need to make another commit to address any standard conventions we decide on in the next day or two.

BKnight760 committed:
Update test monitor swagger doc to be consistent

Fixed the Error object descriptions to be consistent with the other APIs.
Re-ordered the tags, summary, description, operationId, <optional fields>, properties, responses.

BKnight760 committed:
Replace double spaces in sentences with single spaces

Replace double spaces in sentences with single spaces.
